README.md
## catbox-memcached2

[![NPM][npm-image]][npm-url] [![Build Status][travis-image]][travis-url]

Memcached adapter for [catbox](https://github.com/hapijs/catbox).
Supports events, more than 30 days ttl and ready to use for production.

### Installation

```
npm install catbox-memcached2
```

### Usage

#### Options

* `location` - Memcached server location. Default `localhost:11211`
* `partition` - Cache key prefix. Catbox sets partition to `catbox` if it is not
  defined so use `NONE` for no partition.
* `poolSize` - Connection pool size. Default `10`
* `minTimeout` - Connection pool retry min delay before retrying. Default `0`
* `retries` - Connection pool retries to pull connection from pool. Default `0`
* `reconnect` - Reconnect delay after server is dead. Default `1000` ms
* `failures` - Number of failure before marked server as dead. Default `1`
* `retry` - Retry delay when server has an error. Default `1000` ms
* `idle` - Idle time before removing connection from pool. Default `60000` ms
* `flags` - Flags
  * `bypassCacheOnConnError` - Bypass cache if there is a connection error. Default `false`
  * `debug` - List of flags (`['events']`) for debugging. Default `[]`

*See [catbox client](https://github.com/hapijs/catbox#client) and 
[memcached](https://github.com/3rd-Eden/memcached#options) for more details and options.*

### License

Licensed under The MIT License (MIT)  
For the full copyright and license information, please view the LICENSE.txt file.

[npm-url]: http://npmjs.org/package/catbox-memcached2
[npm-image]: https://badge.fury.io/js/catbox-memcached2.png

[travis-url]: https://travis-ci.org/cmfatih/catbox-memcached2
[travis-image]: https://travis-ci.org/cmfatih/catbox-memcached2.svg?branch=master

CHANGELOG.md
## Changelog

### 3.2.0 (2015-07-28)

* Remove duplicate event emitter for get
* Move events upper level
* Update tests
* Add isConnError

### 3.1.0 (2015-07-27)

* Add bypass cache flag

### 3.0.1 (2015-07-24)

* Misc.updates

### 3.0.0 (2015-07-18)

* Change behavior for partition option
* Change cache key structure
* Change events
* Add support for more than 30 days
* Add more tests

### 2.0.0 (2015-07-09)

* Refactor events
* Update readme

### 1.0.2 (2015-07-05)

* Update tests

### 1.0.1 (2015-07-03)

* Upgrade dev dependencies
* Change scripts

### 1.0.0 (2015-05-15)

* Initial release

test/test-lib.js
/* jslint node: true */
/* global describe: false, it: false */
'use strict';

var Client = require('../'),
    expect = require('chai').expect;

// Tests

describe('lib', function() {

  var client;
  beforeEach(function() {
    client = new Client();
  });
  afterEach(function() {
    client.stop();
  });

  it('should create a connection', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);
      done();
    });
  });

  it('should end the connection', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);
      client.stop();
      expect(client.isReady()).to.equal(false);
      done();
    });
  });

  it('should set and get an item', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo', segment: 'test'};
      client.set(key, 'bar', 1000, function(err) {
        expect(err).to.equal(undefined);

        client.get(key, function(err, result) {
          expect(err).to.equal(null);
          expect(result).to.be.a('object');
          expect(result.item).to.equal('bar');
          done();
        });
      });
    });
  });

  it('should set and get an item (30 days)', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo30', segment: 'test'};
      client.drop(key, function(err) {
        expect(err).to.equal(undefined);

        client.set(key, 'bar', 1000*60*60*24*30, function(err) {
          expect(err).to.equal(undefined);

          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.be.a('object');
            expect(result.item).to.equal('bar');
            done();
          });
        });
      });
    });
  });

  it('should set and get an item (>30 days)', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo31', segment: 'test'};
      client.drop(key, function(err) {
        expect(err).to.equal(undefined);

        client.set(key, 'bar', 1000*60*60*24*31, function(err) {
          expect(err).to.equal(undefined);

          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.be.a('object');
            expect(result.item).to.equal('bar');
            done();
          });
        });
      });
    });
  });

  it('should fail to set an item due to value', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'bar', segment: 'test'},
          val = {'baz': 'qux'};
      val.val = val;
      client.set(key, val, 1000, function(err) {
        expect(err.message).to.equal('Converting circular structure to JSON');
        done();
      });
    });
  });

  it('should fail to set an item due to key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.set(null, {}, 1000, function(err) {
        expect(err.message).to.equal('invalid key');

        client.set({}, {}, 1000, function(err) {
          expect(err.message).to.equal('invalid key');
          done();
        });
      });
    });
  });

  it('should fail to set an item due to ttl', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.set({id: 'baz', segment: 'test'}, {}, 'boom', function(err) {
        expect(err.message).to.equal('invalid ttl');
        done();
      });
    });
  });

  it('should return null - key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.get(null, function(err, result) {
        expect(err).to.equal(null);
        expect(result).to.equal(null);
        done();
      });
    });
  });

  it('should fail to return - key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.get({}, function(err, result) {
        expect(err.message).to.equal('invalid key');
        expect(result).to.equal(undefined);
        done();
      });
    });
  });

  it('should drop an item', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'qux', segment: 'test'};
      client.set(key, 'foo', 1, function(err) {
        expect(err).to.equal(undefined);
        client.drop(key, function(err) {
          expect(err).to.equal(undefined);
          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.equal(null);
            done();
          });
        });
      });
    });
  });

  it('should fail to drop an item due to key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.drop(null, function(err, result) {
        expect(err.message).to.equal('invalid key');
        expect(result).to.equal(undefined);

        client.drop({}, function(err, result) {
          expect(err.message).to.equal('invalid key');
          expect(result).to.equal(undefined);
          done();
        });
      });
    });
  });

  it('should validate segment', function(done) {
    expect(client.validateSegmentName().message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\n').message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\0').message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\t').message).to.equal('invalid segment name');
    expect(client.validateSegmentName(' ').message).to.equal('invalid segment name');
    done();
  });

  it('should return null - ttl', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foobar', segment: 'test'};
      client.set(key, 'boom', 1, function(err) {
        expect(err).to.equal(undefined);
        setTimeout(function() {
          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.equal(null);
            done();
          });
        }, 2000);
      });
    });
  });

});

test/test-flags.js
/* jslint node: true */
/* global describe: false, it: false */
'use strict';

var Client = require('../'),
    expect = require('chai').expect;

// Tests

describe('flags', function() {

  var client;
  beforeEach(function() {
    client = new Client({location: '127.0.0.1:11212', 'flags': {'bypassCacheOnConnError': true}});
  });
  afterEach(function() {
    client.stop();
  });

  it('should not fail', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo', segment: 'test'};
      client.set(key, 'bar', 1000, function(err) {
        expect(err).to.equal(undefined);

        client.get(key, function(err, result) {
          expect(err).to.equal(null);
          expect(result).to.equal(null);

          client.drop(key, function(err) {
            expect(err).to.equal(undefined);
            done();
          });
        });
      });
    });
  });

});

test/test-catbox.js
/* jslint node: true */
/* global describe: false, it: false */
'use strict';

var Client = require('../'),
    Catbox = require('catbox'),
    expect = require('chai').expect;

// Tests

describe('catbox', function() {

  var client;
  beforeEach(function() {
    client = new Catbox.Client(Client);
  });
  afterEach(function() {
    client.stop();
  });

  it('should create a connection', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);
      done();
    });
  });

  it('should end the connection', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);
      client.stop();
      expect(client.isReady()).to.equal(false);
      done();
    });
  });

  it('should set and get an item', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo', segment: 'test'};
      client.set(key, 'bar', 1000, function(err) {
        expect(err).to.equal(undefined);

        client.get(key, function(err, result) {
          expect(err).to.equal(null);
          expect(result).to.be.a('object');
          expect(result.item).to.equal('bar');
          done();
        });
      });
    });
  });

  it('should set and get an item (30 days)', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo30', segment: 'test'};
      client.drop(key, function(err) {
        expect(err).to.equal(undefined);

        client.set(key, 'bar', 1000*60*60*24*30, function(err) {
          expect(err).to.equal(undefined);

          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.be.a('object');
            expect(result.item).to.equal('bar');
            done();
          });
        });
      });
    });
  });

  it('should set and get an item (>30 days)', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foo31', segment: 'test'};
      client.drop(key, function(err) {
        expect(err).to.equal(undefined);

        client.set(key, 'bar', 1000*60*60*24*31, function(err) {
          expect(err).to.equal(undefined);

          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.be.a('object');
            expect(result.item).to.equal('bar');
            done();
          });
        });
      });
    });
  });

  it('should fail to set an item due to value', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'bar', segment: 'test'},
          val = {'baz': 'qux'};
      val.val = val;
      client.set(key, val, 1000, function(err) {
        expect(err.message).to.equal('Converting circular structure to JSON');
        done();
      });
    });
  });

  it('should fail to set an item due to key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.set(null, {}, 1000, function(err) {
        expect(err.message).to.equal('Invalid key');

        client.set({}, {}, 1000, function(err) {
          expect(err.message).to.equal('Invalid key');
          done();
        });
      });
    });
  });

  it('should fail to set an item due to ttl', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.set({id: 'baz', segment: 'test'}, {}, 'boom', function(err) {
        expect(err.message).to.equal('invalid ttl');
        done();
      });
    });
  });

  it('should return null - key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.get(null, function(err, result) {
        expect(err).to.equal(null);
        expect(result).to.equal(null);
        done();
      });
    });
  });

  it('should fail to return - key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.get({}, function(err, result) {
        expect(err.message).to.equal('Invalid key');
        expect(result).to.equal(undefined);
        done();
      });
    });
  });

  it('should drop an item', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'qux', segment: 'test'};
      client.set(key, 'foo', 1, function(err) {
        expect(err).to.equal(undefined);
        client.drop(key, function(err) {
          expect(err).to.equal(undefined);
          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.equal(null);
            done();
          });
        });
      });
    });
  });

  it('should fail to drop an item due to key', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      client.drop(null, function(err, result) {
        expect(err.message).to.equal('Invalid key');
        expect(result).to.equal(undefined);

        client.drop({}, function(err, result) {
          expect(err.message).to.equal('Invalid key');
          expect(result).to.equal(undefined);
          done();
        });
      });
    });
  });

  it('should validate segment', function(done) {
    expect(client.validateSegmentName().message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\n').message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\0').message).to.equal('invalid segment name');
    expect(client.validateSegmentName('\t').message).to.equal('invalid segment name');
    expect(client.validateSegmentName(' ').message).to.equal('invalid segment name');
    done();
  });

  it('should return null - ttl', function(done) {
    client.start(function(err) {
      expect(err).to.equal(undefined);
      expect(client.isReady()).to.equal(true);

      var key = {id: 'foobar', segment: 'test'};
      client.set(key, 'boom', 1, function(err) {
        expect(err).to.equal(undefined);
        setTimeout(function() {
          client.get(key, function(err, result) {
            expect(err).to.equal(null);
            expect(result).to.equal(null);
            done();
          });
        }, 2000);
      });
    });
  });

});

lib/index.js
/*
 * catbox-memcached2
 * Copyright (c) 2015 Fatih Cetinkaya (http://github.com/cmfatih/catbox-memcached2)
 * For the full copyright and license information, please view the LICENSE.txt file.
 */

/* jslint node: true */
'use strict';

var Memcached    = require('memcached'),
    util         = require('util'),
    EventEmitter = require('events').EventEmitter;

// Init the module
module.exports = function(options) {

  var serverOptions    = options || {},
      serverLocations  = '127.0.0.1:11211',
      serverPartition, // cache key prefix
      eEmitter,        // event emitter
      flags,           // flags (e.g. `{flags: {debug: ['events']}}`)
      cacheClient;     // cache client

  var TTL_MAX = 1000*60*60*24*30; // 2592000000 milliseconds = 30 days

  // Initialize event emitter
  var EEmitter = function EEmitter() { EventEmitter.call(this); };
  util.inherits(EEmitter, EventEmitter);
  eEmitter = new EEmitter();
  eEmitter.on('error', function() { }); // prevent unhandled error event

  // Server options
  // REF: https://github.com/3rd-Eden/node-memcached/issues/170
  //      https://github.com/3rd-Eden/node-memcached/issues/145#issuecomment-27781592
  //
  // Issues
  // failOverServers: https://github.com/3rd-Eden/memcached/issues/92
  // reconnect/ping : https://github.com/3rd-Eden/memcached/issues/166

  if(serverOptions.location) {
    serverLocations = (serverOptions.location instanceof Array) ? serverOptions.location.join(',') : ''+serverOptions.location;
  }

  if(typeof serverOptions.partition === 'string' && serverOptions.partition !== 'NONE') {
    serverPartition = serverOptions.partition;
  }

  if(!serverOptions.poolSize)     serverOptions.poolSize   = 10;    // default 10
  if(!serverOptions.minTimeout)   serverOptions.minTimeout = 0;     // default 1000
  //if(!serverOptions.maxTimeout) serverOptions.maxTimeout = 1000;  // default 60000
  if(!serverOptions.retries)      serverOptions.retries    = 0;     // default 5
  if(!serverOptions.reconnect)    serverOptions.reconnect  = 1000;  // default 18000000 - reconnect delay after server is dead
  //if(!serverOptions.timeout)    serverOptions.timeout    = 5000;  // default 5000     - connection timeout
  if(!serverOptions.failures)     serverOptions.failures   = 1;     // default 5        - sets server dead
  if(!serverOptions.retry)        serverOptions.retry      = 1000;  // default 30000
  if(!serverOptions.idle)         serverOptions.idle       = 60000; // default 5000
  if(!serverOptions.debug)        serverOptions.debug      = false; // default false

  // Flags
  flags = (serverOptions.flags && typeof serverOptions.flags === 'object' && serverOptions.flags) || {};
  var bypassCacheOnConnError = !!flags.bypassCacheOnConnError;

  // Creates a connection to the cache server
  var start = function start(callback) {

    if(typeof callback !== 'function') {
      throw new Error('invalid callback function');
    }

    // If there is an instance then
    if(cacheClient) {
      return callback();
    }

    // Create instance
    cacheClient = new Memcached(serverLocations, serverOptions);

    // Set events
    cacheClient.on('issue', function(details) {
      eEmitter.emit('error', util.format('issue on %s', serverLocations, details));
    });
    cacheClient.on('failure', function(details) {
      eEmitter.emit('error', util.format('failure on %s', serverLocations, details));
    });
    cacheClient.on('remove', function(details) {
      eEmitter.emit('error', util.format('removing %s', serverLocations, details));
    });
    cacheClient.on('reconnecting', function(details) {
      eEmitter.emit('info', util.format('reconnecting to %s', serverLocations, details));
    });
    cacheClient.on('reconnected', function(details) {
      eEmitter.emit('info', util.format('reconnected to %s', serverLocations, details));
    });
      cacheClient.on('reconnect', function(details) {
      eEmitter.emit('info', util.format('reconnect to %s', serverLocations, details));
    });
    eEmitter.emit('start', util.format('started for %s', serverLocations));

    return callback();
  };

  // Ends the cache server connection
  var stop = function stop() {
    if(cacheClient) {
      cacheClient.end();  // end connection
      cacheClient = null; // cleanup

      eEmitter.emit('stop', util.format('stopped for %s', serverLocations));
    }
  };

  // Returns whether the cache server is ready or not
  var isReady = function isReady() {
    return (!!cacheClient);
  };

  // Validates segment name
  var validateSegmentName = function validateSegmentName(name) {
    if(!name || name.indexOf('\n') !== -1 || name.indexOf('\0') !== -1 || name.indexOf('\t') !== -1 || name.indexOf(' ') !== -1) {
      return new Error('invalid segment name');
    }
    return null; // catbox wants null
  };

  // Gets a value for the given key
  var get = function get(key, callback) {

    if(typeof callback !== 'function') {
      throw new Error('invalid callback function');
    }
    else if(!isReady()) {
      return callback(new Error('cache server is not ready'));
    }

    // If key is falsy then
    if(!key) {
      return callback(null, null); // return null like catbox
    }

    // Get cache value
    var keyF = generateKey(key);

    if(!keyF) {
      return callback(new Error('invalid key'));
    }

    eEmitter.emit('get', keyF);

    cacheClient.get(keyF, function(err, result) {
      if(err) {
        if(bypassCacheOnConnError && isConnError(err)) {
          return callback(null, null); // bypass cache
        }
        return callback(err);
      }

      // No result then no cache
      if(!result) {
        return callback(null, null);
      }

      // Parse cache content
      var envelope = null;
      try {
        envelope = JSON.parse(result);
      }
      catch(e) {
        eEmitter.emit('error', e);
        return callback(e);
      }

      // Check cache content
      if(!envelope || !envelope.item || !envelope.stored) {
        return callback(new Error('invalid cache content'));
      }

      return callback(null, envelope);
    });
  };

  // Sets a value for the given key
  var set = function set(key, value, ttl, callback) {

    if(typeof callback !== 'function') {
      throw new Error('invalid callback function');
    }
    else if(!key) {
      return callback(new Error('invalid key'));
    }
    else if(typeof ttl !== 'number') {
      return callback(new Error('invalid ttl'));
    }
    else if(!isReady()) {
      return callback(new Error('cache server is not ready'));
    }

    // Prepare cache content
    var envelope = {item: value, stored: Date.now(), ttl: ttl},
        envelopeStr;

    try {
      envelopeStr = JSON.stringify(envelope);
    }
    catch(e) {
      eEmitter.emit('error', e);
      return callback(e);
    }

    // Set cache
    var keyF = generateKey(key);

    if(!keyF) {
      return callback(new Error('invalid key'));
    }

    eEmitter.emit('set', keyF);

    // Expiry
    var expiry = Math.max(1, Math.floor(envelope.ttl / 1000)); // min is 1 sec
    // If ttl is greater than max ttl
    if(envelope.ttl > TTL_MAX) {
      // REF: https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L79
      // Calculate timestamp
      expiry = Math.max(1, Math.floor((new Date(envelope.stored + envelope.ttl).getTime()) / 1000)); // min is 1 sec
    }

    cacheClient.set(keyF, envelopeStr, expiry, function(err) {
      if(err) {
        if(bypassCacheOnConnError && isConnError(err)) {
          return callback(); // bypass cache
        }
        return callback(err);
      }

      return callback();
    });
  };

  // Remove the given key
  var drop = function drop(key, callback) {

    if(typeof callback !== 'function') {
      throw new Error('invalid callback function');
    }
    else if(!isReady()) {
      return callback(new Error('cache server is not ready'));
    }

    // Drop cache
    var keyF = generateKey(key);

    if(!keyF) {
      return callback(new Error('invalid key'));
    }

    eEmitter.emit('drop', keyF);

    cacheClient.del(keyF, function(err) {
      if(err) {
        if(bypassCacheOnConnError && isConnError(err)) {
          return callback(); // bypass cache
        }
        return callback(err);
      }

      return callback();
    });
  };

  // Generates keys
  var generateKey = function generateKey(key) {
    return (serverPartition ? encodeURI(serverPartition)+':' : '') + (key && key.segment ? encodeURI(key.segment) : '') + (key && key.id ? ':'+encodeURI(key.id) : '');
  };

  // Allows hook on events
  var on = function on(event, callback) {
    eEmitter.addListener(event, callback);
  };

  // Checks connection error
  var isConnError = function isConnError(err) {
    return (err && (err.code === 'ECONNREFUSED' || (typeof err.message === 'string' && err.message.indexOf('not available') !== -1)));
  };

  // Debug
  //flags = {debug: ['events']}; // for debug
  if(flags.debug instanceof Array) {
    // If events in debug flags
    if(flags.debug.indexOf('events') !== -1) {
      // Log all the events
      ['start', 'stop', 'get', 'set', 'drop', 'info', 'error'].forEach(function(event) {
        on(event, function(data) {
          console.log('\nDebug: ' + JSON.stringify({event: event, data:data}) + '\n');
        });
      });
    }
  }

  // Return
  return {
    start:               start,
    stop:                stop,
    isReady:             isReady,
    validateSegmentName: validateSegmentName,
    get:                 get,
    set:                 set,
    drop:                drop,
    generateKey:         generateKey,
    on:                  on
  };
};